### DISCUSSION TITLE
file uploads using pocketbase as a framework

### DISCUSSION THREAD

**Discussion by eliassebastian** [pocketbase/pocketbase#2489] [Q&A] [answered]
Hi there,

So I've created an additional api endpoint using pocketbase as a framework (e.g. /api/upload). The endpoint then creates a new record within a base collection. One of the fields is set as a file type. The content-type of the request body is multipart/form-data. Haven't really found anything in the documentation about uploading a file when using pocketbase as a framework. Apologies in advance if I've missed something with regards to the documentation. 

Thanks for the help!

**Comment by ganigeorgiev** [comment, marked as answer]
The easiest way is to use `forms.RecordUpsert` struct. You can find an example in https://pocketbase.io/docs/record-methods/#create-new-record-with-data-validations. It shows uploading a local file, but you can use [`filesystem.NewFileFromMultipart`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.15.3/tools/filesystem#NewFileFromMultipart) for multipart/form-data, something like:

```go
e.Router.POST("/upload", func(c echo.Context) error {
	collection, err := app.Dao().FindCollectionByNameOrId("example")
	if err != nil {
		return err
	}

	record := models.NewRecord(collection)

	form := forms.NewRecordUpsert(app, record)

	// load the entire request
	form.LoadRequest(c.Request(), "")

	// ----------------------------------
	// OR manually populate the form data
	// ----------------------------------

	// regular fields
	form.LoadData(map[string]any{
		"someField": 123,
	})

	// uploaded files
	_, mh, err := c.Request().FormFile("yourFileField")
	if err != nil {
		return apis.NewBadRequestError("failed to read form data", err)
	}

	f, err := filesystem.NewFileFromMultipart(mh)
	if err != nil {
		return apis.NewBadRequestError("failed to resolve multipart file", err)
	}

	form.AddFiles("yourFileField", f)

	if err := form.Submit(); err != nil {
		return apis.NewBadRequestError("failed to create the record", err)
	}

	return nil
})
```

**Reply by eliassebastian** [reply]
Thank you! Wasn't aware of the NewFileFromMultipart func. Appreciate the help!

**Reply by vanevalverdes** [reply]
Hi! I saw this answer, but I need do with a JS Hooks, using pocketbase like a framework. How can I do?

Thanks!

**Reply by ganigeorgiev** [reply]
@vanevalverdes Here is a similar JS version of the above (_I haven't tested it_):

```js
routerAdd("POST", "/upload", (c) => {
    const collection = $app.dao().findCollectionByNameOrId("example")

    const record = new Record(collection)

    const form = new RecordUpsertForm($app, record)

    // load the entire request
    form.loadRequest(c.request(), "")

    // ----------------------------------
    // OR manually populate the form data
    // ----------------------------------

    // regular fields
    form.loadData({
        "someField": 123,
    })

    // multipart uploaded files
    const [multipartFile, multipartHeader] = c.request().formFile("yourFileField")

    form.addFiles("yourFileField", $filesystem.fileFromMultipart(multipartHeader))

    try {
        form.submit()
    } catch (err) {
        throw new BadRequestError("failed to create record", err)
    }

    return c.json(200, { success: true })
})
```
_You can also check https://pocketbase.io/docs/js-records/#create-new-record-with-data-validations._

**Reply by vanevalverdes** [reply]
Hi, 

Thanks @ganigeorgiev this solutions works perfect for me.
